### DISCUSSION TITLE
Disable data cache when working with graphql (Apollo Client)

### DISCUSSION THREAD

**Discussion by isabel-sh** [vercel/next.js#48191] [Help]
### Summary

My Next.js project caches data collected from my GraphQL API. This is a massive issue, because the data changes and the caches take priority over the actual result being served by the API. I have searched for what felt like hours trying to find a way to disable the caching in Next.js and have absolutely nothing. My solution is to delete the `.next/cache/fetch-cache` directory to force it to rebuild the cache. The Apollo Client offers the ability to disable caching but this seems to be unrelated to the issue at hand.

How can I disable this feature? It is causing major chafing between me and the framework.

**Comment by alainpaulinn** [comment]
I have the same issue.
Despite having set the following

`
const client = new ApolloClient({
  uri: "https://URL/graphql",
  cache: new InMemoryCache(),
  defaultOptions: {
    watchQuery: {
      fetchPolicy: 'no-cache',
      errorPolicy: 'all',
    },
    query: {
      fetchPolicy: 'no-cache',
      errorPolicy: 'all',
    }
  }
});

export async function getAboutPageContent() {
  const { data } = await client.query({
    query: GET_ALL_CATEGORIES,
    fetchPolicy: 'no-cache'
  });
  return data.categories.data
}
`

Caching is still persisting. My API has constantly changing Data and I need to display the changes

please advise

**Comment by moonstar-x** [comment]
Hey, hope I'm not too late to the party.

This is mostly an issue with the configuration for the Apollo client. Mainly, this is due to the underlying `fetch` that Apollo does in the background.

For regular HTTP calls you can disable Next.js caching with:

```js
fetch(uri, { next: { revalidate: 0 } }
```

The idea is to be able to use this from Apollo.

You can achieve this by passing in a `HttpLink` to the `ApolloClient` constructor. If you originally passed the `uri` and/or `headers` to the `ApolloClient` constructor you will need to move them to the `HttpLink` constructor.

Here's an example:

```ts
import { ApolloClient, InMemoryCache, NormalizedCacheObject, HttpLink, ApolloLink } from '@apollo/client'

const httpLink = new HttpLink({
  uri: `${process.env.GRAPH_QL_URI}/graphql`,
  fetch: function (uri, options) {
    return fetch(uri, {
      ...options ?? {},
      headers: {
        ...options?.headers ?? {},
        Authorization: `Bearer ${process.env.AUTH_TOKEN}`
      },
      next: {
        revalidate: 0
      }
    })
  }
})

export const client: ApolloClient<NormalizedCacheObject> = new ApolloClient({
  cache: new InMemoryCache(),
  link: ApolloLink.from([httpLink]),
  defaultOptions: {
    query: {
      fetchPolicy: 'no-cache',
      errorPolicy: 'all'
    },
    watchQuery: {
      fetchPolicy: 'no-cache',
      errorPolicy: 'all'
    }
  }
})
```

In this case, the `defaultOptions` ensure that Apollo doesn't cache queries in its own cache, and the `next: { revalidate: 0 }` option above tells Next.js to not cache the result of the fetch.

Hope this is of any help.
